src/lib.rs

```diff
@@ -78,14 +75,14 @@ fn do_init_cache(
     supported_features: Buffer,
     cache_size: usize,
 ) -> Result<*mut CosmCache<DB, GoApi, GoQuerier>, Error> {
-    let dir = unsafe { data_dir.read() }.ok_or_else(|| empty_err(DATA_DIR_ARG))?;
-    let dir_str = from_utf8(dir).context(Utf8Err {})?;
+    let dir = unsafe { data_dir.read() }.ok_or_else(|| Error::make_empty_arg(DATA_DIR_ARG))?;
+    let dir_str = from_utf8(dir).map_err(Error::make_invalid_utf8)?;
     // parse the supported features
     let features_bin =
-        unsafe { supported_features.read() }.ok_or_else(|| empty_err(FEATURES_ARG))?;
-    let features_str = from_utf8(features_bin).context(Utf8Err {})?;
+        unsafe { supported_features.read() }.ok_or_else(|| Error::make_empty_arg(FEATURES_ARG))?;
+    let features_str = from_utf8(features_bin).map_err(Error::make_invalid_utf8)?;
```

Comment:
I'd add a `From` conversion for [Utf8Error](https://doc.rust-lang.org/std/str/struct.Utf8Error.html)

Reply:
Done in 2ca001daf158e7469be1e465c275b2e9ab2779ad